python/lsst/daf/butler/tests/cliLogTestBase.py

```diff
@@ -168,6 +171,69 @@ def test_helpLogReset(self):
 
         self.runTest(partial(self.runner.invoke, butlerCli, ["command-log-settings-test", "--help"]))
 
+    def testLongLog(self):
+        """Verify the timestamp is in the log messages when the --long-log
+        flag is set."""
+
+        # When longlog=True, loglines start with the log level and a
+        # timestamp with the following format:
+        # "year-month-day T hour-minute-second.millisecond-zoneoffset"
+        # If lsst.log is importable then the timestamp will have
+        # milliseconds, as described above. If lsst.log is NOT
+        # importable then milliseconds (and the preceding ".") are
+        # omitted (the python `time` module does not support
+        # milliseconds in its format string). Examples of expected
+        # strings follow:
+        # lsst.log:   "DEBUG 2020-10-29T10:20:31.518-0700 ..."
+        # pure python "DEBUG 2020-10-28T10:20:31-0700 ...""
+        # The log level name can change, we verify there is an all
+        # caps word there but do not verify the word. We do not verify
+        # the rest of the log string, assume that if the timestamp is
+        # in the string that the rest of the string will appear as
+        # expected.
+        # N.B. this test is defined in daf_butler which does not depend
+        # on lsst.log. However, CliLog may be used in packages that do
+        # depend on lsst.log and so both forms of timestamps must be
+        # supported. These packages should have a test (the file is
+        # usually called test_cliLog.py) that subclasses CliLogTestBase
+        # and unittest.TestCase so that these tests are run in that
+        # package.
+        timestampRegex = re.compile(
+            r".*[A-Z]+ [0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(.[0-9]{3})?([-,+][0-9]{4}|Z) .*")
+
+        # When longlog=False, log lines start with the module name and
+        # log level, for example:
+        # lsst.daf.butler.core.config DEBUG: ...
+        modulesRegex = re.compile(
+            r".* ([a-z]+\.)+[a-z]+ [A-Z]+: .*")
+
+        with self.runner.isolated_filesystem():
+            for longlog in (True, False):
+                # The click test does not capture logging emitted from lsst.log
+                # so use subprocess to run the test instead.
+                if longlog:
+                    args = ("butler", "--log-level", "DEBUG", "--long-log", "create", "here")
+                else:
+                    args = ("butler", "--log-level", "DEBUG", "create", "here")
+                result = subprocess.run(args, capture_output=True)
```

Comment:
Does the normal click runner not work with lsst.log redirected to python logging? Maybe because one of the tests is using C++ logging? Or is it because click really is forking a subprocess and so logging can never be caught?

Reply:
Click does fork. It sets up its own isolation environment, see https://github.com/pallets/click/blob/master/src/click/testing.py#L172.
There's something about changing the logging configuration, or the way that we do it, that is incompatible with the way click does its isolation environment. I worked on it for a little while but didn't figure out much more than that, and using subprocess as a workaround for the log test seemed better than spending a ton of time on it. FWIW I've verified click.testing does capture python logging output if you don't mess with the log config.
If you want I can dig in more(?)

Reply:
I mainly wanted to know if you had tried the suggested context manager that redirects lsst.log to python logging.

Reply:
`lsstLog.usePythongLogging()` does allow click to capture the log output. Now, without using subprocesses I'm faced with the problem of re-initizliaing the logging system...

Reply:
ok, more like I thought I remembered: logging even with `usePythongLogging` does not get captured by click.

Reply:
surely python logging is always going to be captured by click? You mean the real problem is that click test runner drops log messages completely and doesn't capture them in any way? Okay, that is very funny.

Reply:
When the default python log handler is used python logging gets captured by the click test runner. When we add the `lsst.log.LogHandler` to the python logger then no logging is captured by click, whether we call `lsst.log.Log.usePythonLogging()` or not. I can investigate it more if you want(?) but Fabio has poked me about getting this into the weekly so he can use it, so I'd like to do it on a different ticket. (let me know if you do want?)

Reply:
Ok. Ignore this issue then. You have a solution and you understand what the problem is.